### ISSUE TITLE
Overlay networks in Docker 1.12 and swarm mode

### ISSUE THREAD

**Issue by mschirrmeister** [docker/docker#25295]
Hi,

In the old Docker/Swarm we created overlay networks that the containers for a specific app/service are isolated and could not reach containers from other apps/services.

In Docker 1.12 swarm mode it seems like this is no longer possible.
If I create a service and put it in a overlay network lets say "mynet" and expose a port, the service gets also added to the "ingress" overlay network. If I understand it correct this ingress overlay network (10.255.0.0/16) is used for the load balancing.
All containers in the swarm cluster that have exposed ports via the service can reach each other via this ip from the ingress network.

So, does it mean if I create services and publish ports where the ingress network is used for load balancing, then it does not make sense to assign them also to another overlay network?
And does it also mean the isolation of containers is gone when using the load balancing feature (ingress overlay network) in docker 1.12?

Thanks
Marco

**Output of `docker version`:**

```
Client:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   8eab29e
 Built:
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 2
 Running: 2
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 1.12.0
Storage Driver: btrfs
 Build Version: Btrfs v3.19.1
 Library Version: 101
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: active
 NodeID: d7oq3rjt5llc47hr9wt19tood
 Is Manager: true
 ClusterID: 51zzdq5p2xe8otuwmbalyfy2t
 Managers: 3
 Nodes: 3
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot interval: 10000
  Heartbeat tick: 1
  Election tick: 3
 Dispatcher:
  Heartbeat period: 5 seconds
 CA configuration:
  Expiry duration: 3 months
 Node Address: 10.218.3.5
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 3.10.0-327.22.2.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 6.806 GiB
Name: azeausdockerapps301t.azr.omg.wpp
ID: LWMY:RHUH:JJ5O:OP6G:5LV5:7P7B:WI3W:2JMI:B7HY:EP6J:A7SW:DUX2
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Currently running on Microsoft Azure

**Additional information you deem important (e.g. issue happens only occasionally):**

The above question is more to understand the networking piece and I am not claiming that this is a bug. I assume the current implementation is by design.

**Comment by sanimej** [comment]
@mschirrmeister A service is connected to the ingress network only if it publishes a port. The intent of publishing a port is to make the service available to its client from inside or outside the cluster. If you want a service to be reachable only from within the network (ie: isolated) you shouldn't publish a port. In that case only the containers on the same network can access the service.

In pre 1.12 releases also if you publish a port in docker run containers on other networks can access the service using the host IP and published port.

**Comment by mschirrmeister** [comment]
Thanks for you answer @sanimej 

What you described with pre 1.12 is exactly what we did. We put each service in its own overlay network and the services were only reachable via the exposed ports on the hosts. 
Which means a container of serviceA could never reach or see a container of serviceB.

We did that so that different departments could create isolated servers where their actual containers were never visible/reachable to other departments.
More a separation to avoid questions like, "why are there other containers in my network".

This is gone with 1.12, since all containers are in the ingress network if ports are published.
Again not a technical problem, more that our separation is gone. We will rethink our approach of separating things and will adopt to how 1.12 works.

**Comment by manast** [comment]
you are using expose and publish interchangeably here so I am confused about what is the real status of this. An EXPOSE port should only be accessible by other containers in the same overlay network, thats what I have learnt and assumed is still valid. A published port -p is accessible outside of the containers. Are these assumptions not valid on 1.12 anymore?

**Comment by BretFisher** [comment]
I'm still not clear @mschirrmeister on when you're trying to protect port and when you're trying to publish them. Hopefully this will help clear things up, and if not, maybe a specific example with `service create` and `network create` commands will help us visualize what you want to do:

**TL;DR:** Overlay networks are used for containers to talk to each other on a private network, if you're publishing a port (which will make it available outside the swarm) then I it's expected behavior for any container to reach that port. If you assign each "app/department" their own overlay network and only `expose` ports inside that overlay, then nothing will be seen outside it.

`expose` and `publish` are similar concepts to classic pre-1.12 swarm, so yes be sure you are specific in your wording. 

## Our port options for `service create` are different then `docker run`:

```
root# docker service create --help | grep port
  -p, --publish value               Publish a port as a node port (default [])
root# docker run --help | grep port
      --expose value                Expose a port or a range of ports (default [])
      --health-retries int          Consecutive failures needed to report unhealthy
  -p, --publish value               Publish a container's port(s) to the host (default [])
  -P, --publish-all                 Publish all exposed ports to random ports
```

## So since we can't expose ports in `service create` cli, our two options are:

- `EXPOSE` in Dockerfile should only expose ports inside a swarm network
- `--publish` makes that port available outside swarm, via ingress overlay network

## Observed behaviors:

- If I create a service without specifying an overlay network with `--network` or publishing ports with `--publish`, it'll end up on the default bridge network named `bridge` that has a default route through it's eth0 to exit the swarm. It can see any exposed ports of other containers on that same `bridge` network, but only on that single host, since bridge is not an overlay network.
- If I scale that same service to 2, both containers will still be on the default `bridge` network, but likely now on separate hosts.
- If I 
- Overlay networks are not a default gateway to exit the swarm, so anytime you assign an overlay network to a container you'll see a second interface connected to `docker_gwbridge` which is just there for default route. It can't be used to access any non-published ports on other docker networks.
- If I create a new overlay network named `overlay1` and create two services with `--network overlay1` on it, they can use their service names to access each other over that network.
- If I add a `--publish 80:80` to the above service on `overlay1` then that port will be exposed for incoming traffic to the host on the ingress overlay network, which will route port 80 to the any available container of that service. Those containers still route outbound traffic via `docker_gwbridge`.
- I don't need a customer overlay network if I just need to have a service with multiple replicas of a container published to external. Overlay is just for containers in the swarm to talk to other containers privately (and optionally encrypted).

> We put each service in its own overlay network and the services were only reachable via the exposed (**do you mean published?**) ports on the hosts. Which means a container of serviceA could never reach or see a container of serviceB. <....> This is gone with 1.12, since all containers are in the ingress network if ports are published.

I don't see how anything is gone. So if you `--publish` a port it's exposed to physical network on all nodes in swarm via Routing Mesh... just like "classic" Swarm, but they would be exposed on the node they were running on only.

> all containers are in the ingress network if ports are published.

If by 'in the ingress network' you are suggesting they are connected to it like a bridge or overlay network, I don't believe that to be true. You won't see it in `ip addr`. If they publish ports, they will receive traffic incoming over the ingress network that uses Routing Mesh to get packets to their bridge/overlay interface.

**Comment by xxradar** [comment]
hi ,
I'm not sure if related but ... when creating services like:

docker service create --constraint node.role==worker /
--constraint node.labels.service==stress /
--network stress-net-3 /
--name stress_ab /
--replicas 5 /
dockersec/ab -c 1 -n 1000 http://10.90.122.101/

every-time the container exists, a new container is lauched (as expected).

Although the service is linked to a swarm global overlay, it seems to try todo a IPAM request for INGRESS network 

time="2016-11-02T23:03:44.796307364Z" level=error msg="task allocation failure" error="failed during networktask allocation for task 7qeighnmzapdtx7lk6iyjokww: failed to allocate network IP for task 7qeighnmzapdtx7lk6iyjokww network **3kwq0v051aim4ff2vzwmp1rlh**: could not allocate IP from IPAM: Address already in use" 

docker network ls
**3kwq0v051aim**        ingress             overlay             swarm

**Comment by sheerun** [comment]
@xxradar Maybe you could create new issue about this behavior

**Comment by xxradar** [comment]
@sheerun done.

**Comment by sanimej** [comment]
@mschirrmeister   `This is gone with 1.12, since all containers are in the ingress network if ports are published.`
This is correct. But the tasks belonging to different services were also able to connect to any other (that is not published but the app listens on) port through the ingress network interface. In 1.13 [this is blocked](https://github.com/docker/libnetwork/pull/1496). 

`Again not a technical problem, more that our separation is gone. We will rethink our approach of separating things and will adopt to how 1.12 works.`
Yes, I would suggest treating the `ingress` network differently because its more of a infra network meant for the routing mesh and not an app specific segmented network.

**Comment by aboch** [comment]
As @sanimej explained, docker 1.13.0 now forbids bypassing the network isolation via ingress network (the original problem reported here).

I am going ahead and close this issue then as fixed in 1.13.x by docker/docker/pull/28019

**Comment by mschirrmeister** [comment]
This is great! Confirmed isolation is working again. Example below. Thanks again, great work! 

Create network and services
s1 is in net1
s2 is in net2

```
docker network create --driver overlay net1
docker network create --driver overlay net2

docker service create --name s1 --network net1 -p 8000 jwilder/whoami
docker service create --name s2 --network net2 -p 8000 jwilder/whoami

docker service scale s1=3
docker service scale s2=3
```

Get ingress VIP

```
docker service inspect --format='{{json .Endpoint.VirtualIPs}}' s1
[{"NetworkID":"0oc4ypqnuj6mqad6b0k08zgej","Addr":"10.255.0.4/16"},{"NetworkID":"cswaf831fhnpqrbvfuzxn7tf4","Addr":"10.0.1.2/24"}]

docker service inspect --format='{{json .Endpoint.VirtualIPs}}' s2
[{"NetworkID":"0oc4ypqnuj6mqad6b0k08zgej","Addr":"10.255.0.6/16"},{"NetworkID":"14ztxh8mchmjs9xauo9lakacc","Addr":"10.0.0.2/24"}]
```
Access services on ingress vip from within a container

**1.12.6**

```
s1 to s1 (should work)
root@3649afbc9eca:/app# curl http://10.255.0.4:8000
I'm 3649afbc9eca
s1 to s2 (should not work)
root@3649afbc9eca:/app# curl http://10.255.0.6:8000
I'm 4a8efa04d5bb

s2 to s1 (should not work)
root@e6eba48bab11:/app# curl http://10.255.0.4:8000
I'm 882b1dfbde1a
s2 to s2 (should work)
root@e6eba48bab11:/app# curl http://10.255.0.6:8000
I'm e6eba48bab11
```

**1.13.0**

```
s1 to s1 (should work)
root@2780f37de9d8:/app# curl --connect-timeout 2 http://10.255.0.4:8000
I'm 2780f37de9d8
s1 to s2 (should not work)
root@2780f37de9d8:/app# curl --connect-timeout 2 http://10.255.0.6:8000
curl: (28) Connection timed out after 2001 milliseconds

s2 to s1 (should not work)
root@4cc6db9bf739:/app# curl --connect-timeout 2 http://10.255.0.4:8000
curl: (28) Connection timed out after 2000 milliseconds
s2 to s2 (should work)
root@4cc6db9bf739:/app# curl --connect-timeout 2 http://10.255.0.6:8000
I'm 4cc6db9bf739
```
